pkg/platform/kube/functionres/lazy.go

```diff
@@ -570,6 +607,106 @@ func (lc *lazyClient) waitFunctionDeploymentReadiness(ctx context.Context,
+	return lc.kubeClientSet.AppsV1().
+		Deployments(function.Namespace).
+		Get(ctx, kube.DeploymentNameFromFunctionName(function.Name), metav1.GetOptions{})
+}
+
+func (lc *lazyClient) getFunctionPods(ctx context.Context,
+	function *nuclioio.NuclioFunction) (*v1.PodList, error) {
+	labelSelector := fmt.Sprintf("%s=%s", common.NuclioResourceLabelKeyFunctionName, function.Name)
+	if functionPods, err := lc.kubeClientSet.CoreV1().Pods(function.Namespace).List(ctx,
+		metav1.ListOptions{
+			LabelSelector: labelSelector,
+		},
+	); err == nil {
+		return functionPods, nil
+	} else {
+		return nil, errors.Wrap(err, "Failed to get function deployment's pods")
+	}
+}
+
+// checkFunctionInitContainersDone checks that all function init containers are in terminated status
+// returns (IsDone, reasonNotDone, error)
+// Each pair explanation:
+// (true, "", nil) - all init containers are terminated with 0 exit code, we can proceed to other checks
+// (false, notReadyReason, nil) - some init containers are still waiting/running OR required number of pods hasn't been created yet, so need to wait
+// (false, "", err) - we can stop waiting here since something is broken, so function won't be successfully started
+// (true, err) - impossible
+func (lc *lazyClient) checkFunctionInitContainersDone(ctx context.Context, function *nuclioio.NuclioFunction) (bool, string, error) {
+	functionDeployment, err := lc.getFunctionDeployment(ctx, function)
+	notReadyReason := ""
+	if err != nil {
+		return false, "", err
+	}
+	// deployment doesn't exist yet
+	if functionDeployment == nil {
+		notReadyReason = "deployment doesn't exist yet"
+		return false, notReadyReason, nil
+	}
+
+	// if initContainers aren't defined or replicas number is equal to 0, skip
+	if len(function.Spec.InitContainers) == 0 || functionDeployment.Spec.Replicas == nil {
+		return true, "", nil
+	}
+
+	functionPods, err := lc.getFunctionPods(ctx, function)
+	if err != nil {
+		return false, "", err
+	}
+	// since we are here, it means that we have already checked that the expected number of pods isn't zero
+	// so at least one is expected
+	if functionPods == nil {
+		return false, "", errors.New("No any pods found")
+	}
+
+	// checking that the number of pods is equal to expected replicas, otherwise checking init container
+	// statuses doesn't make sense; need to wait more time
+	if *functionDeployment.Spec.Replicas != int32(len(functionPods.Items)) {
+		notReadyReason = fmt.Sprintf("Not all pod replicas are deployed yet. Expected replicas: %d. Actual replicas: %d",
+			*functionDeployment.Spec.Replicas,
+			len(functionPods.Items))
+		return false, notReadyReason, nil
+	}
+
+	// going through each pod's init containers and check that they all were terminated with exit code 0
+	for _, pod := range functionPods.Items {
+		for _, initContainer := range pod.Status.InitContainerStatuses {
```

Comment:
```suggestion
		for _, initContainerStatus := range pod.Status.InitContainerStatuses {
```

Reply:
@TomerShor I'd leave as is since we can get name/uid/etc from initContainer entity, so it is not really only a status

```
initContainer.Name
```

Reply:
Well, the objects are `ContainerStatus`, although they have a extra fields.
```
// ContainerStatus contains details for the current status of this container.
```

Reply:
renamed